### PR TITLE
Made arguments to specific DebeziumPayloads non-optional, at the cost…

### PR DESCRIPTION
**Pull request by paul-snively**
… of apply being flat-out unsafe.

### CODE REVIEW

**Comment by LukaJCB** [comment]
I'm not sure this is a good idea, we should definitely not use `.get`, but I'm wondering if this is gonna break something. According to this https://debezium.io/documentation/reference/1.2/connectors/postgresql.html#postgresql-create-events 
the `before` and `after` fields are both optional in the schema and the reason we changed them to `Option` in the first place was because I believe @pedrofurla encountered those issues. Maybe we should publish a snapshot and try it out?

**Comment by LukaJCB** [comment]
For reference this is the PR making those changes: https://github.com/compstak/circe-debezium/pull/9/files

**Comment by paul-snively** [comment]
I asked @pedrofurla about that, and he didn't recall such issues. I'm frankly losing patience with Debezium and the whole Kafka Connect ecosystem's incoherence and am leaning towards treating occurrences of it as catastrophic failures to report as bugs against the relevant components. In other words, we expect _our_ intended behavior as the source of truth, and require vendors to articulate and defend their disagreement, if any (and we should be understanding and forgiving in cases where the feedback is simply "we didn't think of that," because we've all been there).

**Comment by paul-snively** [comment]
> According to this https://debezium.io/documentation/reference/1.2/connectors/postgresql.html#postgresql-create-events
> the `before` and `after` fields are both optional in the schema...

Isn't that only because their presence or absence is entirely determined by `op`, and isn't that exactly what we're modeling?

Apart from that, I agree that `apply` should use something like `require` vs. naked `get`s, and if we can agree on the rest, I'll revise the PR accordingly.

**Comment by LukaJCB** [comment]
Yeah let's try this out! I think just converting the `None`s to `Left`s should be fine? Fingers crossed that this works and pedro just made a mistake

File: build.sbt [hunk 1]
```diff
@@ -52,16 +52,34 @@ lazy val publishSettings = Seq(
...
+    Defaults.itSettings,
     name := "circe-debezium",
     libraryDependencies ++= Seq(
       "io.circe" %% "circe-core" % circeV,
-      "io.circe" %% "circe-parser" % circeV % Test,
-      "io.circe" %% "circe-literal" % circeV % Test,
-      "org.scalatest" %% "scalatest" % "3.1.2" % Test
+      "io.circe" %% "circe-parser" % circeV % "test, it",
+      "io.circe" %% "circe-literal" % circeV % "test, it",
+      "ch.qos.logback" % "logback-classic" % "1.2.3" % "runtime",
```
Comment (LukaJCB) [inline build.sbt:70]: could this be `it` scope?

File: src/it/scala/CirceDebeziumSpec.scala [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+  lazy val consumerSettings = ConsumerSettings[IO, DebeziumKey[Int], Option[DebeziumValue[Debezium]]]
+    .withAllowAutoCreateTopics(false)
+    .withAutoOffsetReset(AutoOffsetReset.Earliest)
+    .withBootstrapServers(kafka.bootstrapServers)
+    .withGroupId("group")
+
+  "Interacting with Debezium" - {
+    "DebeziumPayload constructs correctly from real Debezium JSON" in {
+//      forAll (genDoobieProg) { doobie =>
```
Comment (LukaJCB) [inline src/it/scala/CirceDebeziumSpec.scala:198]: How come this doesn't work?

File: build.sbt (hunk 1 above)
Comment (paul-snively) [reply build.sbt:70]: Maybe. I usually leave it as runtime to account for possible transitive dependencies that are also at runtime scope.

File: src/it/scala/CirceDebeziumSpec.scala (hunk 2 above)
Comment (paul-snively) [reply src/it/scala/CirceDebeziumSpec.scala:198]: It probably does. Do we want to try it? It will make the test take a long time if it does work.

Comment (LukaJCB) [reply src/it/scala/CirceDebeziumSpec.scala:198]: Is that just because it creates so many? We can change the default settings of that

Comment (paul-snively) [reply src/it/scala/CirceDebeziumSpec.scala]: OK, I reinstated it, with a minor refactoring to create the connector outside the loop. I don't know that it takes too long. Do you want to try it?

File: src/it/scala/CirceDebeziumSpec.scala [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+  implicit val dbDecoder = Decoder.forProduct8("field1", "field2", "field3", "field4", "field5", "field6", "field7", "field8")(Debezium.apply)
+
+  implicit val intKeyDeserializer = Deserializer.lift { bytes =>
+    IO.fromEither(decode[DebeziumKey[Int]](new String(bytes, "UTF8")))
+  }
+
+  implicit val dbvValueDeserializer = Deserializer.lift { bytes =>
+    IO.fromEither(decode[DebeziumValue[Debezium]](new String(bytes, "UTF8")))
+  }
```
Comment (pedrofurla) [inline src/it/scala/CirceDebeziumSpec.scala:59]: I don't quite understand why you need to do `new String(bytes, "UTF8")`, in the consumers I wrote we got the Json string straight from Kafka. What am I missing?

File: src/it/scala/CirceDebeziumSpec.scala [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  lazy val flyway = Flyway.configure().dataSource(jdbcUrl, pgUser, pgPW).load()
+
+  // A transactor that gets connections from java.sql.DriverManager and executes blocking operations
+  // on an our synchronous EC. See the chapter on connection handling for more info.
+  lazy val transactor = Transactor.fromDriverManager[IO](
+    "org.postgresql.Driver",   // driver classname
+    jdbcUrl,                   // connect URL (driver-specific)
+    pgUser,                // user
+    pgPW,                // password
+    Blocker.liftExecutionContext(ExecutionContexts.synchronous) // just for testing
```
Comment (pedrofurla) [inline src/it/scala/CirceDebeziumSpec.scala:113]: I've see ExecutionContexts.synchronous hang Circle CI. I tried to find a why to "shutdown" but didn't succeed. Ended up using the global EC to solve the hanging. Just keep it in mind if it hangs Circle.

**Comment by pedrofurla** [comment]
Please rename `failure_is_not_an_option` to `failure_is_uninhabited` :)

**Pull request merged**